### PR TITLE
new recipe backend connection

### PR DESCRIPTION
**Pull request by WilliamLi0201**

### CODE REVIEW

File: app/database/models/recipesModel.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+ */
+async function insertAsCreator(recipe, uid){
+    const result = {insertStatus: 'failed'};
+    await db.transaction(async (transaction) => {
+        try {
+            const ids = await db('recipes')
+                .insert(recipe)
+                .transacting(transaction);
+            const saved_recipe = {
+                uid,
```
Comment (rcherukuru2000) [inline app/database/models/recipesModel.js:17]: Codacy has a fix for the issue: [Expected indentation of 8 spaces but found 16.](https://app.codacy.com/gh/cse110-fa21-group17/cse110-fa21-group17/pullRequest?prid=8445077)
```suggestion
        uid,
```

File: app/routes/dashboard.js [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+router.post('/new_recipe', async function(req, res, next){
+    try {
+        const new_rec = req.body;
+        const uid = req.user.id;
+        await recipesModel.insertAsCreator(new_rec, uid);
+        return res.render('pages/AddRecipeSuccess', {title: 'Add Recipe Successful'});
+    } catch(err){
+        console.error(err);
+        return res.status(500)
+            .json({err, data: 'Unable to add recipe, internal server error'});  }
```
Comment (rcherukuru2000) [inline app/routes/dashboard.js:19]: Codacy has a fix for the issue: [Multiple spaces found before '}'.](https://app.codacy.com/gh/cse110-fa21-group17/cse110-fa21-group17/pullRequest?prid=8445077)
```suggestion
            .json({err, data: 'Unable to add recipe, internal server error'}); }
```

File: app/database/models/recipesModel.js [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+                rid: ids[0],
```
Comment (rcherukuru2000) [inline app/database/models/recipesModel.js:18]: Codacy has a fix for the issue: [Expected indentation of 8 spaces but found 16.](https://app.codacy.com/gh/cse110-fa21-group17/cse110-fa21-group17/pullRequest?prid=8445077)
```suggestion
        rid: ids[0],
```

File: app/routes/dashboard.js [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+router.get('/new_recipe', async function(req, res, next){
+    res.render('pages/newrecipe', {title: 'Create Recipe'});
+});
+
+router.post('/new_recipe', async function(req, res, next){
+    try {
+        const new_rec = req.body;
+        const uid = req.user.id;
+        await recipesModel.insertAsCreator(new_rec, uid);
+        return res.render('pages/AddRecipeSuccess', {title: 'Add Recipe Successful'});
```
Comment (rcherukuru2000) [inline app/routes/dashboard.js:15]: Codacy has a fix for the issue: [Expected indentation of 4 spaces but found 8.](https://app.codacy.com/gh/cse110-fa21-group17/cse110-fa21-group17/pullRequest?prid=8445077)
```suggestion
    return res.render('pages/AddRecipeSuccess', {title: 'Add Recipe Successful'});
```

File: app/database/models/recipesModel.js [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+                is_creator: true,
+            };
+            await db('saved_recipes')
+                .insert(saved_recipe)
+                .transacting(transaction);
+            result.insertStatus = 'success';
+            await transaction.commit();
+        } catch (err) {
+            console.log(err);
```
Comment (rcherukuru2000) [inline app/database/models/recipesModel.js:27]: Codacy has a fix for the issue: [Expected indentation of 6 spaces but found 12.](https://app.codacy.com/gh/cse110-fa21-group17/cse110-fa21-group17/pullRequest?prid=8445077)
```suggestion
      console.log(err);
```

File: app/routes/dashboard.js [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+    } catch(err){
+        console.error(err);
```
Comment (rcherukuru2000) [inline app/routes/dashboard.js:17]: Codacy has a fix for the issue: [Expected indentation of 4 spaces but found 8.](https://app.codacy.com/gh/cse110-fa21-group17/cse110-fa21-group17/pullRequest?prid=8445077)
```suggestion
    console.error(err);
```

File: app/database/models/recipesModel.js [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+    const result = {insertStatus: 'failed'};
+    await db.transaction(async (transaction) => {
+        try {
+            const ids = await db('recipes')
+                .insert(recipe)
+                .transacting(transaction);
+            const saved_recipe = {
+                uid,
+                rid: ids[0],
+                is_creator: true,
```
Comment (rcherukuru2000) [inline app/database/models/recipesModel.js:19]: Codacy has a fix for the issue: [Expected indentation of 8 spaces but found 16.](https://app.codacy.com/gh/cse110-fa21-group17/cse110-fa21-group17/pullRequest?prid=8445077)
```suggestion
        is_creator: true,
```

File: app/database/models/recipesModel.js [hunk 8]
```diff
@@ -0,0 +1,35 @@
+const db = require('../dbConfig');
+
+/**
+ * insert recipe as a creator
+ * @param recipe
+ * @param uid
+ * @returns {Promise<{insertStatus: string}>}
+ */
+async function insertAsCreator(recipe, uid){
+    const result = {insertStatus: 'failed'};
```
Comment (rcherukuru2000) [inline app/database/models/recipesModel.js:10]: Codacy has a fix for the issue: [Expected indentation of 2 spaces but found 4.](https://app.codacy.com/gh/cse110-fa21-group17/cse110-fa21-group17/pullRequest?prid=8445077)
```suggestion
  const result = {insertStatus: 'failed'};
```

File: app/database/models/recipesModel.js [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+            await db('saved_recipes')
+                .insert(saved_recipe)
+                .transacting(transaction);
+            result.insertStatus = 'success';
+            await transaction.commit();
+        } catch (err) {
+            console.log(err);
+            result.insertStatus = 'failed';
+            await transaction.rollback();
+        }
```
Comment (rcherukuru2000) [inline app/database/models/recipesModel.js:30]: Codacy has a fix for the issue: [Expected indentation of 4 spaces but found 8.](https://app.codacy.com/gh/cse110-fa21-group17/cse110-fa21-group17/pullRequest?prid=8445077)
```suggestion
    }
```

File: app/database/models/recipesModel.js [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+    });
+    return result;
```
Comment (rcherukuru2000) [inline app/database/models/recipesModel.js:32]: Codacy has a fix for the issue: [Expected indentation of 2 spaces but found 4.](https://app.codacy.com/gh/cse110-fa21-group17/cse110-fa21-group17/pullRequest?prid=8445077)
```suggestion
  return result;
```

File: app/routes/dashboard.js [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+
+/* GET users listing. */
+router.get('/new_recipe', async function(req, res, next){
+    res.render('pages/newrecipe', {title: 'Create Recipe'});
+});
+
+router.post('/new_recipe', async function(req, res, next){
+    try {
+        const new_rec = req.body;
+        const uid = req.user.id;
```
Comment (rcherukuru2000) [inline app/routes/dashboard.js:13]: Codacy has a fix for the issue: [Expected indentation of 4 spaces but found 8.](https://app.codacy.com/gh/cse110-fa21-group17/cse110-fa21-group17/pullRequest?prid=8445077)
```suggestion
    const uid = req.user.id;
```

File: app/database/models/recipesModel.js [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+ * insert recipe as a creator
+ * @param recipe
+ * @param uid
+ * @returns {Promise<{insertStatus: string}>}
+ */
+async function insertAsCreator(recipe, uid){
+    const result = {insertStatus: 'failed'};
+    await db.transaction(async (transaction) => {
+        try {
+            const ids = await db('recipes')
```
Comment (rcherukuru2000) [inline app/database/models/recipesModel.js:13]: Codacy has a fix for the issue: [Expected indentation of 6 spaces but found 12.](https://app.codacy.com/gh/cse110-fa21-group17/cse110-fa21-group17/pullRequest?prid=8445077)
```suggestion
      const ids = await db('recipes')
```

File: app/routes/dashboard.js [hunk 13]
```diff
@@ -0,0 +1,22 @@
+const express = require('express');
+const router = express.Router();
+const recipesModel = require('../database/models/recipesModel');
+
+/* GET users listing. */
+router.get('/new_recipe', async function(req, res, next){
+    res.render('pages/newrecipe', {title: 'Create Recipe'});
```
Comment (rcherukuru2000) [inline app/routes/dashboard.js:7]: Codacy has a fix for the issue: [Expected indentation of 2 spaces but found 4.](https://app.codacy.com/gh/cse110-fa21-group17/cse110-fa21-group17/pullRequest?prid=8445077)
```suggestion
  res.render('pages/newrecipe', {title: 'Create Recipe'});
```

File: app/database/models/recipesModel.js [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+                .insert(recipe)
+                .transacting(transaction);
+            const saved_recipe = {
+                uid,
+                rid: ids[0],
+                is_creator: true,
+            };
```
Comment (rcherukuru2000) [inline app/database/models/recipesModel.js:20]: Codacy has a fix for the issue: [Expected indentation of 6 spaces but found 12.](https://app.codacy.com/gh/cse110-fa21-group17/cse110-fa21-group17/pullRequest?prid=8445077)
```suggestion
      };
```

File: app/database/models/recipesModel.js [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+
+/**
+ * insert recipe as a creator
+ * @param recipe
+ * @param uid
+ * @returns {Promise<{insertStatus: string}>}
+ */
+async function insertAsCreator(recipe, uid){
+    const result = {insertStatus: 'failed'};
+    await db.transaction(async (transaction) => {
```
Comment (rcherukuru2000) [inline app/database/models/recipesModel.js:11]: Codacy has a fix for the issue: [Expected indentation of 2 spaces but found 4.](https://app.codacy.com/gh/cse110-fa21-group17/cse110-fa21-group17/pullRequest?prid=8445077)
```suggestion
  await db.transaction(async (transaction) => {
```

File: app/database/models/recipesModel.js [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+        try {
+            const ids = await db('recipes')
+                .insert(recipe)
+                .transacting(transaction);
```
Comment (rcherukuru2000) [inline app/database/models/recipesModel.js:15]: Codacy has a fix for the issue: [Expected indentation of 10 spaces but found 16.](https://app.codacy.com/gh/cse110-fa21-group17/cse110-fa21-group17/pullRequest?prid=8445077)
```suggestion
          .transacting(transaction);
```

File: app/routes/dashboard.js [hunk 17]
```diff
@@ -0,0 +1,22 @@
...
+
+router.post('/new_recipe', async function(req, res, next){
+    try {
+        const new_rec = req.body;
+        const uid = req.user.id;
+        await recipesModel.insertAsCreator(new_rec, uid);
+        return res.render('pages/AddRecipeSuccess', {title: 'Add Recipe Successful'});
+    } catch(err){
+        console.error(err);
+        return res.status(500)
```
Comment (rcherukuru2000) [inline app/routes/dashboard.js:18]: Codacy has a fix for the issue: [Expected indentation of 4 spaces but found 8.](https://app.codacy.com/gh/cse110-fa21-group17/cse110-fa21-group17/pullRequest?prid=8445077)
```suggestion
    return res.status(500)
```

File: app/routes/dashboard.js (hunk 2 above)
Comment (rcherukuru2000) [inline app/routes/dashboard.js:19]: Codacy has a fix for the issue: [Expected indentation of 8 spaces but found 12.](https://app.codacy.com/gh/cse110-fa21-group17/cse110-fa21-group17/pullRequest?prid=8445077)
```suggestion
        .json({err, data: 'Unable to add recipe, internal server error'});  }
```

File: app/routes/dashboard.js [hunk 18]
```diff
@@ -0,0 +1,22 @@
...
+/* GET users listing. */
+router.get('/new_recipe', async function(req, res, next){
+    res.render('pages/newrecipe', {title: 'Create Recipe'});
+});
+
+router.post('/new_recipe', async function(req, res, next){
+    try {
+        const new_rec = req.body;
+        const uid = req.user.id;
+        await recipesModel.insertAsCreator(new_rec, uid);
```
Comment (rcherukuru2000) [inline app/routes/dashboard.js:14]: Codacy has a fix for the issue: [Expected indentation of 4 spaces but found 8.](https://app.codacy.com/gh/cse110-fa21-group17/cse110-fa21-group17/pullRequest?prid=8445077)
```suggestion
    await recipesModel.insertAsCreator(new_rec, uid);
```

**Review by ThomasLiARDJAVA — APPROVED** [review]

File: app/database/models/recipesModel.js [hunk 19]
```diff
@@ -12,4 +12,37 @@ async function getById(id){
     return data;
 }
 
-module.exports={getById};
+/**
+ * insert recipe as a creator
+ * @param recipe
+ * @param uid
+ * @returns {Promise<{insertStatus: string}>}
```
Comment (rcherukuru2000) [inline app/database/models/recipesModel.js:19]: Codacy has a fix for the issue: [Expected indentation of 8 spaces but found 16.](https://app.codacy.com/gh/cse110-fa21-group17/cse110-fa21-group17/pullRequest?prid=8445077)
```suggestion
        is_creator: true,
```

File: app/routes/dashboard.js (hunk 2 above)
Comment (rcherukuru2000) [inline app/routes/dashboard.js:19]: Codacy has a fix for the issue: [Unexpected space(s) before '}'.](https://app.codacy.com/gh/cse110-fa21-group17/cse110-fa21-group17/pullRequest?prid=8445077)
```suggestion
            .json({err, data: 'Unable to add recipe, internal server error'});}
```

File: app/routes/dashboard.js (hunk 17 above)
Comment (rcherukuru2000) [inline app/routes/dashboard.js:18]: Codacy has a fix for the issue: [Expected indentation of 4 spaces but found 8.](https://app.codacy.com/gh/cse110-fa21-group17/cse110-fa21-group17/pullRequest?prid=8445077)
```suggestion
    return res.status(500)
```

File: app/routes/dashboard.js (hunk 6 above)
Comment (rcherukuru2000) [inline app/routes/dashboard.js:17]: Codacy has a fix for the issue: [Expected indentation of 4 spaces but found 8.](https://app.codacy.com/gh/cse110-fa21-group17/cse110-fa21-group17/pullRequest?prid=8445077)
```suggestion
    console.error(err);
```

File: app/database/models/recipesModel.js [hunk 20]
```diff
@@ -12,4 +12,37 @@ async function getById(id){
...
+ */
+async function insertAsCreator(recipe, uid){
+    const result = {insertStatus: 'failed'};
```
Comment (rcherukuru2000) [inline app/database/models/recipesModel.js:22]: Codacy has a fix for the issue: [Expected indentation of 10 spaces but found 16.](https://app.codacy.com/gh/cse110-fa21-group17/cse110-fa21-group17/pullRequest?prid=8445077)
```suggestion
          .insert(saved_recipe)
```

File: app/database/models/recipesModel.js [hunk 21]
```diff
@@ -12,4 +12,37 @@ async function getById(id){
     return data;
 }
 
-module.exports={getById};
+/**
+ * insert recipe as a creator
+ * @param recipe
```
Comment (rcherukuru2000) [inline app/database/models/recipesModel.js:17]: Codacy has a fix for the issue: [Expected indentation of 8 spaces but found 16.](https://app.codacy.com/gh/cse110-fa21-group17/cse110-fa21-group17/pullRequest?prid=8445077)
```suggestion
        uid,
```

File: app/database/models/recipesModel.js [hunk 22]
```diff
@@ -12,4 +12,37 @@ async function getById(id){
...
+ * @param uid
+ * @returns {Promise<{insertStatus: string}>}
+ */
+async function insertAsCreator(recipe, uid){
+    const result = {insertStatus: 'failed'};
+    await db.transaction(async (transaction) => {
+        try {
+            const ids = await db('recipes')
+                .insert(recipe)
```
Comment (rcherukuru2000) [inline app/database/models/recipesModel.js:26]: Codacy has a fix for the issue: [Expected indentation of 4 spaces but found 8.](https://app.codacy.com/gh/cse110-fa21-group17/cse110-fa21-group17/pullRequest?prid=8445077)
```suggestion
    } catch (err) {
```

File: app/database/models/recipesModel.js [hunk 23]
```diff
@@ -12,4 +12,37 @@ async function getById(id){
...
+                .transacting(transaction);
+            const saved_recipe = {
```
Comment (rcherukuru2000) [inline app/database/models/recipesModel.js:28]: Codacy has a fix for the issue: [Expected indentation of 6 spaces but found 12.](https://app.codacy.com/gh/cse110-fa21-group17/cse110-fa21-group17/pullRequest?prid=8445077)
```suggestion
      result.insertStatus = 'failed';
```

File: app/database/models/recipesModel.js [hunk 24]
```diff
@@ -12,4 +12,37 @@ async function getById(id){
...
+                uid,
```
Comment (rcherukuru2000) [inline app/database/models/recipesModel.js:29]: Codacy has a fix for the issue: [Expected indentation of 6 spaces but found 12.](https://app.codacy.com/gh/cse110-fa21-group17/cse110-fa21-group17/pullRequest?prid=8445077)
```suggestion
      await transaction.rollback();
```

File: app/database/models/recipesModel.js [hunk 25]
```diff
@@ -12,4 +12,37 @@ async function getById(id){
     return data;
 }
 
-module.exports={getById};
+/**
+ * insert recipe as a creator
+ * @param recipe
+ * @param uid
+ * @returns {Promise<{insertStatus: string}>}
+ */
```
Comment (rcherukuru2000) [inline app/database/models/recipesModel.js:20]: Codacy has a fix for the issue: [Expected indentation of 6 spaces but found 12.](https://app.codacy.com/gh/cse110-fa21-group17/cse110-fa21-group17/pullRequest?prid=8445077)
```suggestion
      };
```

File: app/routes/dashboard.js (hunk 11 above)
Comment (rcherukuru2000) [inline app/routes/dashboard.js:13]: Codacy has a fix for the issue: [Expected indentation of 4 spaces but found 8.](https://app.codacy.com/gh/cse110-fa21-group17/cse110-fa21-group17/pullRequest?prid=8445077)
```suggestion
    const uid = req.user.id;
```

File: app/database/models/recipesModel.js [hunk 26]
```diff
@@ -12,4 +12,37 @@ async function getById(id){
...
+async function insertAsCreator(recipe, uid){
+    const result = {insertStatus: 'failed'};
+    await db.transaction(async (transaction) => {
+        try {
+            const ids = await db('recipes')
+                .insert(recipe)
+                .transacting(transaction);
```
Comment (rcherukuru2000) [inline app/database/models/recipesModel.js:27]: Codacy has a fix for the issue: [Expected indentation of 6 spaces but found 12.](https://app.codacy.com/gh/cse110-fa21-group17/cse110-fa21-group17/pullRequest?prid=8445077)
```suggestion
      console.log(err);
```

File: app/routes/dashboard.js (hunk 2 above)
Comment (rcherukuru2000) [inline app/routes/dashboard.js:19]: Codacy has a fix for the issue: [Multiple spaces found before '}'.](https://app.codacy.com/gh/cse110-fa21-group17/cse110-fa21-group17/pullRequest?prid=8445077)
```suggestion
            .json({err, data: 'Unable to add recipe, internal server error'}); }
```

File: app/database/models/recipesModel.js [hunk 27]
```diff
@@ -12,4 +12,37 @@ async function getById(id){
...
+/**
+ * insert recipe as a creator
+ * @param recipe
+ * @param uid
+ * @returns {Promise<{insertStatus: string}>}
+ */
+async function insertAsCreator(recipe, uid){
+    const result = {insertStatus: 'failed'};
+    await db.transaction(async (transaction) => {
+        try {
```
Comment (rcherukuru2000) [inline app/database/models/recipesModel.js:24]: Codacy has a fix for the issue: [Expected indentation of 6 spaces but found 12.](https://app.codacy.com/gh/cse110-fa21-group17/cse110-fa21-group17/pullRequest?prid=8445077)
```suggestion
      result.insertStatus = 'success';
```

File: app/database/models/recipesModel.js [hunk 28]
```diff
@@ -12,4 +12,37 @@ async function getById(id){
     return data;
 }
 
-module.exports={getById};
+/**
```
Comment (rcherukuru2000) [inline app/database/models/recipesModel.js:15]: Codacy has a fix for the issue: [Expected indentation of 10 spaces but found 16.](https://app.codacy.com/gh/cse110-fa21-group17/cse110-fa21-group17/pullRequest?prid=8445077)
```suggestion
          .transacting(transaction);
```

File: app/database/models/recipesModel.js [hunk 29]
```diff
@@ -12,4 +12,37 @@ async function getById(id){
...
+ * insert recipe as a creator
+ * @param recipe
+ * @param uid
+ * @returns {Promise<{insertStatus: string}>}
+ */
+async function insertAsCreator(recipe, uid){
+    const result = {insertStatus: 'failed'};
+    await db.transaction(async (transaction) => {
+        try {
+            const ids = await db('recipes')
```
Comment (rcherukuru2000) [inline app/database/models/recipesModel.js:25]: Codacy has a fix for the issue: [Expected indentation of 6 spaces but found 12.](https://app.codacy.com/gh/cse110-fa21-group17/cse110-fa21-group17/pullRequest?prid=8445077)
```suggestion
      await transaction.commit();
```

File: app/routes/dashboard.js (hunk 2 above)
Comment (rcherukuru2000) [inline app/routes/dashboard.js:19]: Codacy has a fix for the issue: [Closing curly brace should be on the same line as opening curly brace or on the line after the previous block.](https://app.codacy.com/gh/cse110-fa21-group17/cse110-fa21-group17/pullRequest?prid=8445077)
```suggestion
            .json({err, data: 'Unable to add recipe, internal server error'});  
```

**Pull request merged**